### PR TITLE
db: config: move consistent-topology-changes out of experimental and make it the default for new clusters

### PR DESCRIPTION
**Pull request by patjed41**
We move consistent cluster management out of experimental and
make it the default for new clusters in 6.0. In code, we make the 
`consistent-topology-changes` flag unused and assumed to be true.

In 6.0, the topology upgrade procedure will be manual and
voluntary, so some clusters will still be using the gossip-based
topology even though they support the raft-based topology.
Therefore, we need to continue testing the gossip-based topology.
This is possible by using the `force-gossip-topology-changes` flag
introduced in scylladb/scylladb#18284.

Ref scylladb/scylladb#17802

### CODE REVIEW

File: service/storage_service.cc [hunk 1]
```diff
@@ -1922,7 +1922,7 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     }
 
     assert(_group0);
-    co_await _group0->finish_setup_after_join(*this, _qp, _migration_manager.local(), _raft_experimental_topology);
+    co_await _group0->finish_setup_after_join(*this, _qp, _migration_manager.local(), false);
```
Comment (kbr-scylla) [inline service/storage_service.cc:1920]: @piodul please review this commit (` storage_service: join_token_ring: fix finish_setup_after_join calls `)

File: test/boost/tablets_test.cc [hunk 2]
```diff
@@ -264,6 +264,10 @@ SEASTAR_TEST_CASE(test_tablet_metadata_persistence) {
 }
 
 SEASTAR_TEST_CASE(test_read_required_hosts) {
+    // FIXME: the test fails without using disable_consistent_topology_changes.
```
Comment (kbr-scylla) [inline test/boost/tablets_test.cc:267]: reminder to open issue

Comment (kbr-scylla) [reply test/boost/tablets_test.cc:267]: so this was using tablets **but not** consistent-topology? super weird

Comment (kbr-scylla) [reply test/boost/tablets_test.cc:267]: hm, maybe enabling consistent topology causes load balancer to start running, and the load balancer preconditions are not satisfied because this test only partially updates the metadata, or something.

So potentially one way to fix it would be to disable the load balancer (IIRC we have some API for that)

Comment (patjed41) [reply test/boost/tablets_test.cc:267]: Yes, normally using tablets and not using raft-topology throws bad config. But this is thrown from the `main` function which is not executed in boost unit tests.

File: test/boost/schema_change_test.cc [hunk 3]
```diff
@@ -779,6 +779,7 @@ future<> test_schema_digest_does_not_change_with_disabled_features(sstring data_
     }
     cql_test_config cfg_in(db_cfg_ptr);
     cfg_in.disabled_features = std::move(disabled_features);
+    cfg_in.run_with_raft_recovery = true;
```
Comment (kbr-scylla) [inline test/boost/schema_change_test.cc:782]: Would be good to explain in comment why we need this -- the explanation in commit message is good but will be hard to find

Comment (kbr-scylla) [inline test/boost/schema_change_test.cc:782]: > Then, it gets stuck in the group 0 upgrade (for a valid reason - messaging to itself is disabled).

Hm, that sounds like a bug in group 0 upgrade. In a single-node cluster it should not depend on messaging. In fact there is some logic in upgrade procedure to address the case of being  run from boost tests, see `wait_for_peers_to_enter_synchronize_state`:
```cpp
    auto entered_synchronize = make_lw_shared<std::unordered_set<gms::inet_address>>();

    // This is a work-around for boost tests where RPC module is not listening so we cannot contact ourselves.
    // But really, except the (arguably broken) test code, we don't need to be treated as an edge case. All nodes are symmetric.
    // For production code this line is unnecessary.
    entered_synchronize->insert(ms.broadcast_address());
```

where does the procedure get stuck?

File: service/storage_service.cc (hunk 1 above)
Comment (piodul) [reply service/storage_service.cc:1920]: Looks good.

I suppose we didn't catch this earlier because that would require doing a rolling upgrade from a version which didn't have a group0 created yet (if we already joined then we won't create the `group0_state_machine` again), but the code after the change is obviously more correct.

**Review by kbr-scylla — COMMENTED** [review]
Only some minor comments left, generally looks good.

**Comment by avikivity** [comment]
I don't understand, is this moving consistent topology to be non-experimental, or making it mandatory?

The ceremonial patch should be sent be @kbr-scylla or @kostja who led the effort.

**Comment by avikivity** [comment]
> This is possible by using the disable-consistent-topology-changes

We should avoid double-negative flags where you have to compute that disable=false actually means enable.

**Comment by kostja** [comment]
> > This is possible by using the disable-consistent-topology-changes
> 
> We should avoid double-negative flags where you have to compute that disable=false actually means enable.

I made the same comment but @kbr-scylla says he's approved the option name with @tzach . 

@tzach could you clarify?

**Comment by kostja** [comment]
> I don't understand, is this moving consistent topology to be non-experimental, or making it mandatory?
> 
> The ceremonial patch should be sent be @kbr-scylla or @kostja who led the effort.

I think in the end we decided there should be a way to disable it to be able to continue testing of non-upgraded clusters, since the upgrade is not mandatory. Hence the option name.

**Comment by patjed41** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8032/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8032/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9027/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8032/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1003/)
> #### Failed Tests (84/30310):


Failures in `gating-dtest-release` are expected because tests that are supposed to run with the gossip-based topology ran with the raft-based topology. PR in scylla-dtest that will be merged before this PR will fix it (PR 2. in https://github.com/scylladb/scylladb/issues/17802#issuecomment-2056836938).

There is one unexpected failure in `gating-dtest-release-with-consistent-topology-changes`, in `system_tables_test.TestConfigTable.test_content_source_config`. This test verifies that `scylla.yaml` matches the `system.config` table. If the test is run with `consistent-topology-changes`, `scylla.yaml` contains it and `system.config` doesn't, so the test fails. The solution could be:
- disable this test case (or disable only checking `experimental_features`) in PR 2. in https://github.com/scylladb/scylladb/issues/17802#issuecomment-2056836938,
- reenable this test case in PR 4. -- this PR will also stop passing `consistent-topology-changes`, so the test should start passing again.

**Comment by avikivity** [comment]
> * it makes consistent topology non-experimental (removes the experimental feature flag)

I read the title, it says it makes it unused.

And again, I think @kostja or you should send this last patch.

**Comment by kbr-scylla** [comment]
> I read the title, it says it makes it unused.

Removing an experimental flag is implemented in code as marking it `UNUSED`. It's an implementation detail. Same thing we did when we made `cdc` or `lwt` experimental.
```cpp
std::map<sstring, db::experimental_features_t::feature> db::experimental_features_t::map() {
    // We decided against using the construct-on-first-use idiom here:
    // https://github.com/scylladb/scylla/pull/5369#discussion_r353614807
    // Features which are no longer experimental are mapped
    // to UNUSED switch for a while, and can be eventually
    // removed altogether.
    return {
        {"lwt", feature::UNUSED},
        {"udf", feature::UDF},
        {"cdc", feature::UNUSED},
```

> And again, I think @kostja or you should send this last patch.

I have my team for that.
I'm also not sure what you mean by "last patch".
The last PR will go into scylla-dtest repo as explained in this plan: https://github.com/scylladb/scylladb/issues/17802#issuecomment-2056836938

File: test/boost/schema_change_test.cc (hunk 3 above)
Comment (patjed41) [reply test/boost/schema_change_test.cc:782]: It gets stuck in
```cpp
future<> raft_group0::do_upgrade_to_group0(group0_upgrade_state start_state, service::storage_service& ss, cql3::query_processor& qp, service::migration_manager& mm, bool topology_change_enabled) {
    ...
    co_await check_remote_group0_upgrade_state_dry_run(get_inet_addrs, _ms.local(), _abort_source);
```
Some relevant logs:
```
INFO  2024-04-18 19:06:45,589 [shard 0:main] storage_service - Restarting in legacy mode. The node was either upgraded from a non-raft-topology version or is restarting after recovery.
...
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0 - finish_setup_after_join: SUPPORTS_RAFT feature enabled. Starting internal upgrade-to-raft procedure.
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - starting in `use_pre_raft_procedures` state.
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - Waiting until everyone is ready to start upgrade...
...
WARN  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - check_remote_group0_upgrade_state_dry_run: `send_get_group0_upgrade_state(127.0.0.1)` failed: seastar::rpc::closed_error (connection is closed)
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - future<> service::check_remote_group0_upgrade_state_dry_run(const noncopyable_function<future<std::vector<gms::inet_address>> ()> &, netw::messaging_service &, abort_source &): sleeping for 1s seconds before retrying...
```

I don't understand why the only change addressing boost tests is in `wait_for_peers_to_enter_synchronize_state`. I guess we should never enter `raft_group0::do_upgrade_to_group0` in boost tests. If yes, then there is nothing to fix. `wait_for_peers_to_enter_synchronize_state` is also called in `raft_group0::setup_group0`. Maybe that's why only this function was changed. 

Anyway, I still think this test should run in RECOVERY mode. This node behaves as if it restarted after majority loss (unless I'm missing something).

Comment (kbr-scylla) [reply test/boost/schema_change_test.cc:782]: > I don't understand why the only change addressing boost tests is in wait_for_peers_to_enter_synchronize_state. I guess we should never enter raft_group0::do_upgrade_to_group0 in boost tests. If yes, then there is nothing to fix. wait_for_peers_to_enter_synchronize_state is also called in raft_group0::setup_group0. Maybe that's why only this function was changed.

Indeed. setup_group0 is prepared for disabled messaging (for boost test) but the full group 0 upgrade procedure is not. The assumption was we'd never enter this procedure in boost test because we'd start nodes with Raft enabled from the beginning, in which case we don't go through the procedure.

File: test/auth_cluster/test_raft_service_levels.py [hunk 4]
```diff
@@ -59,13 +59,13 @@ async def test_service_levels_snapshot(manager: ManagerClient):
 @pytest.mark.asyncio
 @skip_mode('release', 'error injections are not supported in release mode')
```
Comment (piodul) [inline test/auth_cluster/test_raft_service_levels.py:60]: > We prefer a flag over an error injection because we will use it
> a lot in CI jobs' configurations, some tests, manual testing etc.
> It's much more convenient.

Another positive side effect of this is that this flag works in release mode. Please go over the tests you modified and see which ones should be re-enabled in release mode. Look at the `@skip_mode` annotations but also look at `skip_in_release` in `suite.yaml`.

File: test/boost/schema_change_test.cc [hunk 5]
```diff
@@ -779,6 +779,7 @@ future<> test_schema_digest_does_not_change_with_disabled_features(sstring data_
     }
     cql_test_config cfg_in(db_cfg_ptr);
```
Comment (piodul) [inline test/boost/schema_change_test.cc:780]: > In the following commit, we remove the `consistent-topology-changes`
> experimental feature.

Nit: IIUC we don't really remove it in the next commit, we only remove any effect it might have on the code.

**Review by piodul — COMMENTED** [review]
I suppose I will have to review some commits again after the name `disable-consistent-topology-changes` is changed to `force-gossip-topology-changes`... especially that the flag will be flipped.

**Comment by patjed41** [comment]
v2:
- rebased on v2 of #18284
- updated commit messages and cover letter after renaming `disabled-consistent-topology-changes` to `force-gossip-topology-changes`
- re-enabled in the release mode tests that were disabled only because of the `force_gossip_based_join` error injection (3rd commit)
- made booting fail if the node is using `force-gossip-topology-changes` and joining the cluster that is using the raft-based topology (3rd commit); @kbr-scylla suggested this change
- running `test_schema_digest_does_not_change_with_disabled_features` in the raft RECOVERY mode (5th commit):
  - changed explanation in the commit message
  - added a comment with the explanation
- made a change in some commit messages: "remove `consistent-topology-changes`" -> "make `consistent-topology-changes` unused" - we don't really remove the feature, we only remove any effect it might have on the code

File: test/auth_cluster/test_raft_service_levels.py (hunk 4 above)
Comment (patjed41) [reply test/auth_cluster/test_raft_service_levels.py]: Good catch, I did it.

File: test/boost/schema_change_test.cc (hunk 5 above)
Comment (patjed41) [reply test/boost/schema_change_test.cc:781]: I fixed it in this commit and some other commits.

**Comment by patjed41** [comment]
> I suppose I will have to review some commits again after the name `disable-consistent-topology-changes` is changed to `force-gossip-topology-changes`... especially that the flag will be flipped.

Fortunately, the flag is not flipped. It's still false by default.

**Review by kbr-scylla — APPROVED** [review]

**Comment by piodul** [comment]
> > I suppose I will have to review some commits again after the name `disable-consistent-topology-changes` is changed to `force-gossip-topology-changes`... especially that the flag will be flipped.
> 
> Fortunately, the flag is not flipped. It's still false by default.

Right. Removal of the negative from the name confused me...

**Review by piodul — APPROVED** [review]
Looks good, but needs to be rebased first.

**Comment by patjed41** [comment]
Well, this doesn't look good:

> #### Build Details:
> 
>   * Duration: 20 hr

> Identified problems
>
>   Job cancelled due to timeout
>
>   2024-04-23T20:01:21.035Z [63428/65000] topology_custom dev [ PASS ] topology_custom.test_table_drop.28 2.80s
>
>   2024-04-23T20:01:21.268Z Cancelling nested steps due to timeout

```
08:57:27  [60401/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.1 729.50s
08:57:35  [60402/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.2 723.46s
...
13:47:32  [60497/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.97 721.63s
13:47:49  [60498/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.98 723.48s
```

It looks like PRs that make trivial changes to many tests suffer after we started running every changed unit test 100 times. I don't know how big this problem is. I guess PRs like this one are rare.

We also have a test that takes 5 hours in the debug mode...

**Comment by patjed41** [comment]
> #### Failed Tests (1/12129):
> 
>     * [test_recover_stuck_raft_recovery](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8170/testReport/junit/%28root%29/test_raft_recovery_stuck/test_recover_stuck_raft_recovery) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery)

It looks like another instance of #18369.

```
        logging.info(f"Restarting {others}")
        await asyncio.gather(*(restart(manager, srv) for srv in others))
        cql = await reconnect_driver(manager)
    
        logging.info(f"{others} restarted, waiting until driver reconnects to them")
        hosts = await wait_for_cql_and_get_hosts(cql, others, time.time() + 60)
    
        logging.info(f"Checking if {hosts} are in recovery state")
        for host in hosts:
>           rs = await cql.run_async(
                    "select value from system.scylla_local where key = 'group0_upgrade_state'",
                    host=host)
E           cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.2.219.5:9042 datacenter1>: ConnectionException('Host has been marked down or removed')})
```
Just before the failuire:
```
09:38:53.927 DEBUG> Shutting down connections to 127.2.219.5:9042
09:38:53.927 DEBUG> All shards of host 127.2.219.15:9042 have at least one connection, closing 0 excess connections
09:38:53.927 DEBUG> Closing connection (139715555165904) to 127.2.219.5:9042
09:38:53.927 DEBUG> Closing connection (139715555165904) to 127.2.219.5:9042
09:38:53.927 DEBUG> Closed socket to 127.2.219.5:9042
09:38:53.927 DEBUG> Closing pending connection (139715557994896) to 127.2.219.5:9042
09:38:53.927 DEBUG> Closing connection (139715557994896) to 127.2.219.5:9042
09:38:53.927 DEBUG> Closed socket to 127.2.219.5:9042
09:38:53.927 DEBUG> Closing excess connection (139715557994896) to 127.2.219.5:9042
```
Node 5427 uses `127.2.219.5`. Logs of 5428:
```
INFO  2024-04-23 09:38:53,843 [shard 0: gms] gossip - InetAddress 3fa69919-76ac-4979-acd9-63a426093e42/127.2.219.16 is now UP, status = NORMAL
INFO  2024-04-23 09:38:53,920 [shard 0: gms] gossip - InetAddress e5442fac-8dc5-4f1b-9da2-1739da6bfef0/127.2.219.5 is now DOWN, status = shutdown
WARN  2024-04-23 09:38:53,922 [shard 0: gms] gossip - Skip marking node 127.2.219.5 with status = shutdown as UP
------ Ending test debug/topology_custom.test_raft_recovery_stuck.76::test_recover_stuck_raft_recovery ------
INFO  2024-04-23 09:38:54,027 [shard 0: gms] storage_service - handle_state_normal for 127.2.219.5/e5442fac-8dc5-4f1b-9da2-1739da6bfef0 finished
INFO  2024-04-23 09:38:54,042 [shard 0: gms] gossip - InetAddress e5442fac-8dc5-4f1b-9da2-1739da6bfef0/127.2.219.5 is now UP, status = NORMAL
```

**Comment by kbr-scylla** [comment]
https://github.com/scylladb/scylladb/pull/18285#issuecomment-2074291904

@yaronkaikov @benipeled can we get a regular CI run on this PR (just run each test once as normal)?

**Comment by kbr-scylla** [comment]
Actually let's wait until rebase

**Comment by avikivity** [comment]
> db: config: make consistent-topology-changes unused #18285

Please pay more attention to wordings. The title is precise but gives the completely wrong impression. What does it mean that consistent-topology-changes are unused? Are topology changes no longer consistent? In fact it's the opposite.

The series moves consistent topology changes out of experimental mode and makes it the default for new clusters. But that can't be inferred from a casual read.

**Comment by patjed41** [comment]
rebased: the only changes are in the ` storage_service: join_cluster: replace force_gossip_based_join with force-gossip-topology-changes` commit after introducing `set_topology_change_kind`

**Comment by patjed41** [comment]
> > db: config: make consistent-topology-changes unused #18285
> 
> Please pay more attention to wordings. The title is precise but gives the completely wrong impression. What does it mean that consistent-topology-changes are unused? Are topology changes no longer consistent? In fact it's the opposite.
> 
> The series moves consistent topology changes out of experimental mode and makes it the default for new clusters. But that can't be inferred from a casual read.

You are right, I changed the wording.

**Comment by patjed41** [comment]
I also updated the cover letter. It could be confusing for the same reasons the title was.

I also removed the obsolete paragraph about rebase on #18284 and merge order.

**Comment by kbr-scylla** [comment]
> Well, this doesn't look good:
> 
> > #### Build Details:
> > 
> > * Duration: 20 hr
> 
> > Identified problems
> > Job cancelled due to timeout
> > 2024-04-23T20:01:21.035Z [63428/65000] topology_custom dev [ PASS ] topology_custom.test_table_drop.28 2.80s
> > 2024-04-23T20:01:21.268Z Cancelling nested steps due to timeout
> 
> ```
> 08:57:27  [60401/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.1 729.50s
> 08:57:35  [60402/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.2 723.46s
> ...
> 13:47:32  [60497/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.97 721.63s
> 13:47:49  [60498/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.98 723.48s
> ```
> 
> It looks like PRs that make trivial changes to many tests suffer after we started running every changed unit test 100 times. I don't know how big this problem is. I guess PRs like this one are rare.
> 
> We also have a test that takes 5 hours in the debug mode...

1 run takes ~12 minutes, we do 100 runs with concurrency (pool_size) 4, so 25 runs in each "thread", giving us 5h.

Basically any PR which touches this test (`test_tablets_migration`) will have to wait at least 5h for CI.
This is horrible.

@xemul can we do something so this test does not take 12 minutes?

**Comment by kbr-scylla** [comment]
Thanks @yaronkaikov !

**Comment by xemul** [comment]
> > Well, this doesn't look good:
> > > #### Build Details:
> > > 
> > > * Duration: 20 hr
> > 
> > 
> > > Identified problems
> > > Job cancelled due to timeout
> > > 2024-04-23T20:01:21.035Z [63428/65000] topology_custom dev [ PASS ] topology_custom.test_table_drop.28 2.80s
> > > 2024-04-23T20:01:21.268Z Cancelling nested steps due to timeout
> > 
> > 
> > ```
> > 08:57:27  [60401/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.1 729.50s
> > 08:57:35  [60402/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.2 723.46s
> > ...
> > 13:47:32  [60497/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.97 721.63s
> > 13:47:49  [60498/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.98 723.48s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It looks like PRs that make trivial changes to many tests suffer after we started running every changed unit test 100 times. I don't know how big this problem is. I guess PRs like this one are rare.
> > We also have a test that takes 5 hours in the debug mode...
> 
> 1 run takes ~12 minutes, we do 100 runs with concurrency (pool_size) 4, so 25 runs in each "thread", giving us 5h.
> 
> Basically any PR which touches this test (`test_tablets_migration`) will have to wait at least 5h for CI. This is horrible.
> 
> @xemul can we do something so this test does not take 12 minutes?

@kbr-scylla , not much, unfortunately, see #17764 (spoiler: it's 16 tests under one hood there, so less than a minute each)

**Comment by xemul** [comment]
> > @xemul can we do something so this test does not take 12 minutes?
> 
> @kbr-scylla , not much, unfortunately, see #17764 (spoiler: it's 16 tests under one hood there, so less than a minute each)

On the other hand we, probably, can. Quoting the original issue:

> >    I see. What takes time? Work or waiting?
> 
> It's "waiting". Most of the time tests wait for coordinator to catch up with removenode's adding of node to ignored list (~30 sec). Cluster bootstrap is silver winner (~5 sec).

Reducing the removenode catch-up timing would make things notably better

**Comment by yaronkaikov** [comment]
> > > Well, this doesn't look good:
> > > > #### Build Details:
> > > > 
> > > > * Duration: 20 hr
> > > 
> > > 
> > > > Identified problems
> > > > Job cancelled due to timeout
> > > > 2024-04-23T20:01:21.035Z [63428/65000] topology_custom dev [ PASS ] topology_custom.test_table_drop.28 2.80s
> > > > 2024-04-23T20:01:21.268Z Cancelling nested steps due to timeout
> > > 
> > > 
> > > ```
> > > 08:57:27  [60401/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.1 729.50s
> > > 08:57:35  [60402/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.2 723.46s
> > > ...
> > > 13:47:32  [60497/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.97 721.63s
> > > 13:47:49  [60498/65000] topology_custom  debug  [ PASS ] topology_custom.test_tablets_migration.98 723.48s
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It looks like PRs that make trivial changes to many tests suffer after we started running every changed unit test 100 times. I don't know how big this problem is. I guess PRs like this one are rare.
> > > We also have a test that takes 5 hours in the debug mode...
> > 
> > 
> > 1 run takes ~12 minutes, we do 100 runs with concurrency (pool_size) 4, so 25 runs in each "thread", giving us 5h.
> > Basically any PR which touches this test (`test_tablets_migration`) will have to wait at least 5h for CI. This is horrible.
> > @xemul can we do something so this test does not take 12 minutes?
> 
> I have added to the CI an external parameter so we can adjust the number of repeats for specific test cases. So in case we have a PR like this one, we can modify it Running CI again https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8293/

It doesn't make much sense to run those tests only once or any other number. I think that if we have a PR with more than 10 test changes, we should run the CI as we run it today. @kbr-scylla does it make sense?

**Comment by kbr-scylla** [comment]
> It doesn't make much sense to run those tests only once or any other number. I think that if we have a PR with more than 10 test changes, we should run the CI as we run it today. @kbr-scylla does it make sense?

Doesn't  "as we run it today" mean "only once"?

**Comment by yaronkaikov** [comment]
> > It doesn't make much sense to run those tests only once or any other number. I think that if we have a PR with more than 10 test changes, we should run the CI as we run it today. @kbr-scylla does it make sense?
> 
> Doesn't "as we run it today" mean "only once"?

No, it means we run each test 9 times (3 for each mode)

**Comment by kbr-scylla** [comment]
> No, it means we run each test 9 times (3 for each mode)

OK

**Comment by patjed41** [comment]
> #### 🔴 CI State: FAILURE

#16002

**Comment by kbr-scylla** [comment]
> 🔴 CI State: FAILURE

https://github.com/scylladb/scylladb/issues/15192
